File: Data Structures & Algorithms/interleaving-string/submission-5.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        auto back = [&](auto self, int i, int j){
            if (i == s1.size() && j == s2.size()) {
                return true;
            }
            else if (i > s1.size() || j > s2.size()) return false;
            bool res{false};
            if (s1[i] == s3[i + j]) res = res || self(self, i + 1, j);
            if (s2[j] == s3[i + j]) res = res || self(self, i, j + 1);
            return res;
        };
        return back(back, 0, 0);
    }
};

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-5.java
class Solution {
    private final String[] digitToChar = {
            "", "", "abc", "def", "ghi", "jkl",
            "mno", "qprs", "tuv", "wxyz"
        };
    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        StringBuilder cur = new StringBuilder();
        if (digits.length() == 0) return new ArrayList<>();
        back(res, digits, 0, cur);

        return res;
    }
    private void back(List<String> res, String digits, int index, StringBuilder cur) {
        if (index == digits.length()) {
            String combination = new String(cur);
            res.add(combination);
            return;
        }
        int i = (digits.charAt(index) - '0');
        for (char c : digitToChar[i].toCharArray()) {

            back(res, digits, index + 1, new StringBuilder(cur).append(c));
        }
    }
}
